### PR TITLE
Create blobs & directories manually in 'createBlockBlob'

### PR DESCRIPTION
**Pull request by wwlorey**
Fixes https://github.com/microsoft/vscode-azurestorage/issues/601
Fixes https://github.com/microsoft/vscode-azurestorage/issues/574

### CODE REVIEW

File: src/utils/blobUtils.ts [hunk 1]
```diff
@@ -56,26 +56,14 @@ export async function loadMoreBlobChildren(parent: BlobContainerTreeItem | BlobD
...
-    });
+    let blobPath: string = context.childName || await showBlobPathInputBox(parent);
 
     return await vscode.window.withProgress({ location: vscode.ProgressLocation.Window }, async (progress) => {
         context.showCreatingTreeItem(blobPath);
         progress.report({ message: `Azure Storage: Creating block blob '${blobPath}'` });
         await createOrUpdateBlockBlob(parent, blobPath);
         // Use the full blob path as the label since this is a newly created blob. https://github.com/microsoft/vscode-azurestorage/issues/565
-        return new BlobTreeItem(parent, blobPath, parent.container, blobPath);
+        return new BlobTreeItem(parent, blobPath, parent.container);
```
Comment (EricJizbaMSFT) [inline src/utils/blobUtils.ts]: Remove the comment above since it no longer applies. Also just get rid of `label` as a parameter to the constructor in `BlobTreeItem` since I don't think it's ever used

File: src/commands/blob/blobContainerActionHandlers.ts [hunk 2]
```diff
@@ -6,17 +6,46 @@
 import * as vscode from 'vscode';
```
Comment (EricJizbaMSFT) [inline src/commands/blob/blobContainerActionHandlers.ts]: Is this related to 574 or does it fix 574? It seems like it fixes it, in which case please remove the "wont-fix" label and put it in the current milestone

File: src/commands/blob/blobContainerActionHandlers.ts [hunk 3]
```diff
@@ -6,17 +6,46 @@
...
+
+        for (const dirName of dirNames) {
+            currentPath += `${dirName}/`;
+            const dirUri: vscode.Uri = AzureStorageFS.idToUri(parent.fullId, currentPath);
+
+            try {
+                let fileType: vscode.FileType = (await ext.azureStorageFS.stat(dirUri)).type;
+                if (fileType !== vscode.FileType.Directory) {
+                    // Remove trailing slash from currentPath for error message
+                    throw new Error(`${currentPath.slice(0, -1)} is not a directory`);
```
Comment (EricJizbaMSFT) [inline src/commands/blob/blobContainerActionHandlers.ts]: nit: Please get in the habit of using `localize` for new strings

File: src/commands/blob/blobContainerActionHandlers.ts [hunk 4]
```diff
@@ -6,17 +6,46 @@
...
+                }
```
Comment (EricJizbaMSFT) [inline src/commands/blob/blobContainerActionHandlers.ts]: Any time you have a dynamically inserted value in a string, wrap it in quotes

`This is not a directory` vs. `"This" is not a directory`

File: src/commands/blob/blobContainerActionHandlers.ts [hunk 5]
```diff
@@ -6,17 +6,46 @@
...
 
 export function registerBlobContainerActionHandlers(): void {
     registerCommand("azureStorage.openBlobContainer", openBlobContainerInStorageExplorer);
     registerCommand("azureStorage.editBlob", async (_context: IActionContext, treeItem: BlobTreeItem) => AzureStorageFS.showEditor(treeItem), 250);
     registerCommand("azureStorage.deleteBlobContainer", async (context: IActionContext, treeItem?: BlobContainerTreeItem) => await deleteNode(context, BlobContainerTreeItem.contextValue, treeItem));
-    registerCommand("azureStorage.createBlockBlob", async (context: IActionContext, treeItem: BlobContainerTreeItem) => {
-        let childTreeItem = await treeItem.createChild(context);
+    registerCommand("azureStorage.createBlockBlob", async (context: IActionContext, parent: BlobContainerTreeItem) => {
+        const blobPath: string = await showBlobPathInputBox(parent);
+        const dirNames: string[] = blobPath.split('/').slice(0, -1);
```
Comment (EricJizbaMSFT) [inline src/commands/blob/blobContainerActionHandlers.ts]: Are there `path` methods we could use instead? I get nervous about off-by-one errors, etc. whenever I see hard-coded zeros, ones, and slashes

File: src/commands/blob/blobContainerActionHandlers.ts (hunk 2 above)
Comment (wwlorey) [reply src/commands/blob/blobContainerActionHandlers.ts]: I thought to fix 574 it would have to expand the new blob when its created. Right now it just displays the root directory and you have to expand it yourself

Comment (EricJizbaMSFT) [reply src/commands/blob/blobContainerActionHandlers.ts]: Oh I thought the issue was more about getting the correct tree nodes and didn't even think about the expansion state

File: src/utils/blobUtils.ts [hunk 6]
```diff
@@ -119,6 +108,77 @@ export async function getExistingProperties(parent: BlobTreeItem | BlobContainer
...
+                return localize('aBlobWithThisPathAndNameAlreadyExists', 'A blob with this path and name already exists');
+            }
+
+            return undefined;
+        }
+    });
+}
+
+export async function findBlobTreeItem(uri: vscode.Uri, filePath: string, containerTreeItem: BlobContainerTreeItem, context: IActionContext, endSearchEarly?: boolean): Promise<AzureStorageBlobTreeItem> {
+    let treeItem: BlobTreeItem | BlobDirectoryTreeItem | BlobContainerTreeItem = containerTreeItem;
```
Comment (EricJizbaMSFT) [inline src/utils/blobUtils.ts]: I think `ext.tree.findTreeItem` would suffice for your scenario instead of pulling this code out into `findBlobTreeItem`

File: src/utils/blobUtils.ts [hunk 7]
```diff
@@ -119,6 +108,77 @@ export async function getExistingProperties(parent: BlobTreeItem | BlobContainer
...
+
+    return treeItem;
+}
+
+export async function getBlobResourceType(filePath: string, containerTreeItem: BlobContainerTreeItem, context: IActionContext): Promise<vscode.FileType> {
+    let treeItem: AzureStorageBlobTreeItem = await findBlobTreeItem(vscode.Uri.parse(containerTreeItem.fullId), filePath, containerTreeItem, context);
+    return treeItem instanceof BlobDirectoryTreeItem || treeItem instanceof BlobContainerTreeItem ? vscode.FileType.Directory : vscode.FileType.File;
+}
+
+export async function createBlobDirectory(fullDirPath: string, parentPath: string, parentTreeItem: BlobDirectoryTreeItem | BlobContainerTreeItem, context: IActionContext): Promise<void> {
```
Comment (EricJizbaMSFT) [inline src/utils/blobUtils.ts]: Most of the code you pulled out into this function doesn't apply to your scenario. You already know the exact dirName and you're only creating one directory at a time. Just call createChild directly. Something like this:
```
await dirParentTreeItem.createChild(<IBlobContainerCreateChildContext>{ ...context, childType: BlobDirectoryTreeItem.contextValue, childName: dirName });
```
